File: ios/App/App/KickZone/AI/Encoder.swift
// Swift port of js/rl/encoder.js (the 1v1 encoder with frame stacking).
//
// Produces a 56-feature observation per frame from the player's
// perspective. The runtime agent stacks the last 3 frames to feed into
// the network (168 dims).
//
// We keep this exactly compatible with the JS encoder so the Swift
// inference behaves identically to the trained desktop model.

import CoreGraphics

enum RLEncoder {
    static let featureDim = 56
    static let stackK = 3
    static let stackedDim = featureDim * stackK

    /// Encode `self`'s view of the world into `out`. Caller must pass an
    /// array of at least `featureDim` elements (we mutate in place).
    static func encode(self_ s: Player, opp o: Player, ball b: Ball, field f: Field,
                       timeLeftMs: Double, scoreDiff: Int, kickoffActive: Bool,
                       out: inout [Float]) {
        let isRed = (s.team == .red)
        let fw = f.width
        let fh = f.height
        let fx = f.x
        let fy = f.y
        let maxBallSpeed = Float(GameConstants.maxBallSpeed)

        // Side-coord helpers (red passes through; blue mirrors x)
        @inline(__always) func sideX(_ x: CGFloat) -> Float {
            isRed ? Float((x - fx) / fw) : Float((fx + fw - x) / fw)
        }
        @inline(__always) func sideY(_ y: CGFloat) -> Float {
            Float((y - fy) / fh)
        }
        @inline(__always) func sideVx(_ vx: CGFloat) -> Float {
            (isRed ? Float(vx) : -Float(vx)) / maxBallSpeed
        }
        @inline(__always) func sideVy(_ vy: CGFloat) -> Float {
            Float(vy) / maxBallSpeed
        }

        var i = 0
        // --- Self (14)
        out[i] = sideX(s.pos.x) * 2 - 1; i += 1
        out[i] = sideY(s.pos.y) * 2 - 1; i += 1
        out[i] = sideVx(s.vel.x); i += 1
        out[i] = sideVy(s.vel.y); i += 1
        out[i] = Float(min(s.kickCooldownMs / 200, 1)); i += 1
        out[i] = Float(min(s.pullCooldownMs / 8000, 1)); i += 1
        out[i] = s.powerUp == .speed ? 1 : 0; i += 1
        out[i] = s.powerUp == .ghost ? 1 : 0; i += 1
        out[i] = s.powerUp == .shield ? 1 : 0; i += 1
        out[i] = s.powerUp == .dash ? 1 : 0; i += 1
        out[i] = s.powerUp == .frozen ? 1 : 0; i += 1
        out[i] = s.powerUp == .slowed ? 1 : 0; i += 1
        out[i] = s.stunMs > 0 ? 1 : 0; i += 1
        out[i] = s.pullActive ? 1 : 0; i += 1

        // --- Opp (10) — relative
        out[i] = sideX(o.pos.x) - sideX(s.pos.x); i += 1
        out[i] = sideY(o.pos.y) - sideY(s.pos.y); i += 1
        out[i] = sideVx(o.vel.x); i += 1
        out[i] = sideVy(o.vel.y); i += 1
        out[i] = Float(min(o.kickCooldownMs / 200, 1)); i += 1
        out[i] = o.powerUp == .speed ? 1 : 0; i += 1
        out[i] = o.powerUp == .ghost ? 1 : 0; i += 1
        out[i] = o.powerUp == .shield ? 1 : 0; i += 1
        out[i] = o.powerUp == .dash ? 1 : 0; i += 1
        out[i] = (o.stunMs > 0 || o.powerUp == .frozen) ? 1 : 0; i += 1

        // --- Ball (9)
        let bx = sideX(b.pos.x)
        let by = sideY(b.pos.y)
        out[i] = bx - sideX(s.pos.x); i += 1
        out[i] = by - sideY(s.pos.y); i += 1
        out[i] = sideVx(b.vel.x); i += 1
        out[i] = sideVy(b.vel.y); i += 1
        out[i] = b.superKick > 0 ? 1 : 0; i += 1
        out[i] = b.ghost ? 1 : 0; i += 1
        let poss: Float = (b.lastKickedBy === s) ? 1 : (b.lastKickedBy === o ? -1 : 0)
        out[i] = poss; i += 1
        out[i] = bx - sideX(o.pos.x); i += 1
        out[i] = by - sideY(o.pos.y); i += 1

        // --- Predicted ball 15 frames ahead (4)
        let pred = predictBall(b, frames: 15)
        let pbx = sideX(pred.x)
        let pby = sideY(pred.y)
        out[i] = pbx - sideX(s.pos.x); i += 1
        out[i] = pby - sideY(s.pos.y); i += 1
        out[i] = pbx - bx; i += 1
        out[i] = pby - by; i += 1

        // --- Goals (3)
        out[i] = 0 - sideX(s.pos.x); i += 1
        out[i] = 1 - sideX(s.pos.x); i += 1
        let goalCenterY = Float((f.goalY + f.goalHeight / 2 - fy) / fh)
        out[i] = goalCenterY - sideY(s.pos.y); i += 1

        // --- Game state (3)
        out[i] = Float(timeLeftMs / 60000); i += 1
        out[i] = tanhf(Float(scoreDiff) * 0.5); i += 1
        out[i] = kickoffActive ? 1 : 0; i += 1

        // --- Shooting features (8)
        let kickRange = Float(s.radius + b.radius + 21)
        let dRaw = Float(((b.pos.x - s.pos.x) * (b.pos.x - s.pos.x)
                       + (b.pos.y - s.pos.y) * (b.pos.y - s.pos.y)).squareRoot())
        let inKickRange: Float = dRaw < kickRange ? 1 : 0
        out[i] = inKickRange; i += 1
        let kdx = Float(b.pos.x - s.pos.x)
        let kdy = Float(b.pos.y - s.pos.y)
        let klen = max((kdx * kdx + kdy * kdy).squareRoot(), 1)
        let knx = kdx / klen
        let kny = kdy / klen
        let oppGoalX = isRed ? Float(fx + fw) : Float(fx)
        let goalCenterAbsY = Float(f.goalY + f.goalHeight / 2)
        let tgx = oppGoalX - Float(b.pos.x)
        let tgy = goalCenterAbsY - Float(b.pos.y)
        let tglen = max((tgx * tgx + tgy * tgy).squareRoot(), 1)
        let tnx = tgx / tglen
        let tny = tgy / tglen
        let aimDot = knx * tnx + kny * tny
        out[i] = aimDot; i += 1
        out[i] = (aimDot > 0.7 && inKickRange == 1) ? 1 : 0; i += 1
        let perp = abs(kdx * tny - kdy * tnx)
        out[i] = min(perp / Float(fw), 1); i += 1
        out[i] = (isRed ? tgx : -tgx) / Float(fw); i += 1
        out[i] = tgy / Float(fh); i += 1
        let opx = Float(o.pos.x - b.pos.x)
        let opy = Float(o.pos.y - b.pos.y)
        let oprojF = opx * tnx + opy * tny
        let oprojP = abs(opx * tny - opy * tnx)
        let oppInLane: Float = (oprojF > 0 && oprojF < tglen && oprojP < 50) ? 1 : 0
        out[i] = oppInLane; i += 1
        out[i] = min(tglen / Float(fw), 1); i += 1

        // Pad zeros up to featureDim
        while i < featureDim { out[i] = 0; i += 1 }
    }

    private static func predictBall(_ ball: Ball, frames: Int) -> Vec2 {
        var x = ball.pos.x, y = ball.pos.y
        var vx = ball.vel.x, vy = ball.vel.y
        let f = GameConstants.ballFriction
        for _ in 0..<frames {
            vx *= f; vy *= f
            x += vx; y += vy
        }
        return Vec2(x, y)
    }
}

/// Ring-buffer style frame stack.
final class FrameStack {
    let featureDim: Int
    let k: Int
    private var frames: [[Float]] = []
    private(set) var stacked: [Float]

    init(featureDim: Int, k: Int) {
        self.featureDim = featureDim
        self.k = k
        self.stacked = [Float](repeating: 0, count: featureDim * k)
    }

    func fill(_ frame: [Float]) {
        frames = Array(repeating: frame, count: k)
        rebuild()
    }
    func push(_ frame: [Float]) {
        if frames.count == k { frames.removeFirst() }
        frames.append(frame)
        rebuild()
    }
    private func rebuild() {
        for i in 0..<k {
            let base = i * featureDim
            let f = frames[i]
            for j in 0..<featureDim { stacked[base + j] = f[j] }
        }
    }
}

File: js/rl/encoder.js
// State encoder + action decoder for KickZone 1v1 RL.
//
// Goals:
//   - Compact, normalized observation that captures everything the agent needs.
//   - Symmetric: the same encoder works regardless of team (red/blue) by
//     internally mirroring blue's view so the policy only ever learns one frame.
//   - Mirror augmentation across the X axis (vertical mirror) doubles training
//     data for free — the game is symmetric vertically.
//
// Convention: "side" coordinates are always from the agent's perspective.
//   - +x_side = "forward" (toward enemy goal)
//   - +y_side = "down" on the field
// For a red player, side coords == world coords.
// For a blue player, x is flipped (world.x -> field.right - world.x).
(function(root, factory) {
    const Physics = (typeof require !== 'undefined') ? require('../../shared/physics') : root.Physics;
    const exp = factory(Physics);
    if (typeof module !== 'undefined' && module.exports) module.exports = exp;
    else root.RLEncoder = exp;
})(typeof self !== 'undefined' ? self : this, function(Physics) {

const FEATURE_DIM = 56;
// Frame stacking: the policy sees the last STACK_K observations concatenated
// in [oldest, ..., newest] order. This gives the agent implicit access to
// acceleration, ball curve, and short-term opponent intent without an LSTM.
const STACK_K = 3;
const STACKED_DIM = FEATURE_DIM * STACK_K;
const NUM_POWERUP_TYPES = 6; // speed, ghost, dash, shield, freeze, slow
const POWERUP_INDEX = { speed: 0, ghost: 1, dash: 2, shield: 3, freeze: 4, slow: 5 };

// Mirror flag controls whether to flip Y axis (for data augmentation)
function encode(self, opp, ball, field, gameState, powerUps, out, mirrorY = false) {
    const isRed = self.team === 'red';
    const fw = field.width;
    const fh = field.height;
    const fx = field.x;
    const fy = field.y;
    const fcx = field.centerX;
    const fcy = field.centerY;

    // Convert world -> agent-side coords. For blue, flip across centerX.
    const sideX = (x) => isRed ? (x - fx) / fw : (fx + fw - x) / fw; // [0..1]
    const sideY = (y) => {
        const v = (y - fy) / fh;
        return mirrorY ? (1 - v) : v; // [0..1]
    };
    const sideVx = (vx) => (isRed ? vx : -vx) / Physics.MAX_BALL_SPEED;
    const sideVy = (vy) => (mirrorY ? -vy : vy) / Physics.MAX_BALL_SPEED;

    let i = 0;

    // --- Self ---
    out[i++] = sideX(self.x) * 2 - 1;
    out[i++] = sideY(self.y) * 2 - 1;
    out[i++] = sideVx(self.vx);
    out[i++] = sideVy(self.vy);
    out[i++] = Math.min(self.kickCooldown / 200, 1);
    out[i++] = Math.min(self.pullCooldown / 8000, 1);
    // Powerup one-hot (4 main types — others handled via flags)
    out[i++] = self.powerUp === 'speed' ? 1 : 0;
    out[i++] = self.powerUp === 'ghost' ? 1 : 0;
    out[i++] = self.powerUp === 'shield' ? 1 : 0;
    out[i++] = self.powerUp === 'dash' ? 1 : 0;
    out[i++] = (self.powerUp === 'frozen') ? 1 : 0;
    out[i++] = (self.powerUp === 'slowed') ? 1 : 0;
    out[i++] = self.stunTimer > 0 ? 1 : 0;
    out[i++] = self.pullActive ? 1 : 0;

    // --- Opponent (relative to self in side coords) ---
    out[i++] = sideX(opp.x) - sideX(self.x);
    out[i++] = sideY(opp.y) - sideY(self.y);
    out[i++] = sideVx(opp.vx);
    out[i++] = sideVy(opp.vy);
    out[i++] = Math.min(opp.kickCooldown / 200, 1);
    out[i++] = opp.powerUp === 'speed' ? 1 : 0;
    out[i++] = opp.powerUp === 'ghost' ? 1 : 0;
    out[i++] = opp.powerUp === 'shield' ? 1 : 0;
    out[i++] = opp.powerUp === 'dash' ? 1 : 0;
    out[i++] = (opp.stunTimer > 0 || opp.powerUp === 'frozen') ? 1 : 0;

    // --- Ball ---
    const bx = sideX(ball.x);
    const by = sideY(ball.y);
    out[i++] = bx - sideX(self.x);
    out[i++] = by - sideY(self.y);
    out[i++] = sideVx(ball.vx);
    out[i++] = sideVy(ball.vy);
    out[i++] = ball.superKick > 0 ? 1 : 0;
    out[i++] = ball.ghost ? 1 : 0;
    // Possession: 1 = self touched last, -1 = opponent, 0 = neither
    out[i++] = ball.lastKickedBy === self ? 1 : (ball.lastKickedBy === opp ? -1 : 0);
    out[i++] = bx - sideX(opp.x); // ball relative to opponent (helps positioning)
    out[i++] = by - sideY(opp.y);

    // --- Predicted ball (15 frames ahead) ---
    const pred = predictBall(ball, 15);
    const pbx = sideX(pred.x);
    const pby = sideY(pred.y);
    out[i++] = pbx - sideX(self.x);
    out[i++] = pby - sideY(self.y);
    out[i++] = pbx - bx; // ball drift direction
    out[i++] = pby - by;

    // --- Goals (relative side coords) ---
    // Own goal is at sideX=0, opponent goal at sideX=1
    out[i++] = 0 - sideX(self.x); // dist to own goal x
    out[i++] = 1 - sideX(self.x); // dist to opp goal x
    const goalCenterY = (field.goalY + field.goalHeight / 2 - fy) / fh;
    const myGoalY = mirrorY ? (1 - goalCenterY) : goalCenterY;
    out[i++] = myGoalY - sideY(self.y);

    // --- Game state ---
    out[i++] = (gameState.timeLeft || 0) / 60000;
    out[i++] = Math.tanh((gameState.scoreDiff || 0) * 0.5);
    out[i++] = gameState.kickoffActive ? 1 : 0;

    // --- Shooting quality features (8) ---
    // These tell the agent: "If I kicked right now, would it score?"
    // Computing these once per encode is much cheaper than the network learning
    // to derive them from raw positions and velocities.
    const kickRange = self.radius + ball.radius + 21;
    const ballDistRaw = Math.sqrt((ball.x - self.x) * (ball.x - self.x) + (ball.y - self.y) * (ball.y - self.y));
    const inKickRange = ballDistRaw < kickRange;
    out[i++] = inKickRange ? 1 : 0;
    // Kick direction (player -> ball, the direction the ball would go if kicked now)
    const kdx = ball.x - self.x;
    const kdy = ball.y - self.y;
    const klen = Math.max(Math.sqrt(kdx * kdx + kdy * kdy), 1);
    const knx = kdx / klen;
    const kny = kdy / klen;
    // Vector from ball to opponent goal
    const oppGoalX = isRed ? (fx + fw) : fx;
    const goalCenterYAbs = field.goalY + field.goalHeight / 2;
    const tgx = oppGoalX - ball.x;
    const tgy = goalCenterYAbs - ball.y;
    const tglen = Math.max(Math.sqrt(tgx * tgx + tgy * tgy), 1);
    const tnx = tgx / tglen;
    const tny = tgy / tglen;
    // Cosine of angle between kick direction and direction-to-goal: 1 = perfect aim
    const aimDot = knx * tnx + kny * tny;
    out[i++] = aimDot;
    out[i++] = aimDot > 0.7 && inKickRange ? 1 : 0; // "shooting opportunity now"
    // Distance from current kick line to goal center (perpendicular distance)
    // This tells the agent how far off-line their kick would be.
    const perpDist = Math.abs(kdx * tny - kdy * tnx); // cross product magnitude
    out[i++] = Math.min(perpDist / fw, 1);
    // Vector from ball to opponent goal (in side coords)
    out[i++] = (isRed ? tgx : -tgx) / fw;
    out[i++] = (mirrorY ? -tgy : tgy) / fh;
    // Is opponent in the shooting lane? (along ball -> goal line)
    const opx = opp.x - ball.x;
    const opy = opp.y - ball.y;
    const oprojForward = opx * tnx + opy * tny;
    const oprojPerp = Math.abs(opx * tny - opy * tnx);
    const oppInLane = (oprojForward > 0 && oprojForward < tglen && oprojPerp < 50) ? 1 : 0;
    out[i++] = oppInLane;
    out[i++] = Math.min(tglen / fw, 1); // ball-to-goal distance, normalized

    // --- Nearest power-up ---
    let nearestPU = null;
    let nearestPUDist = Infinity;
    if (powerUps && powerUps.length) {
        for (const pu of powerUps) {
            const d = Physics.distance(self, pu);
            if (d < nearestPUDist) { nearestPUDist = d; nearestPU = pu; }
        }
    }
    if (nearestPU) {
        out[i++] = sideX(nearestPU.x) - sideX(self.x);
        out[i++] = sideY(nearestPU.y) - sideY(self.y);
        out[i++] = 1; // present
        const pid = (nearestPU.type && nearestPU.type.id) || nearestPU.id;
        const idx = POWERUP_INDEX[pid] !== undefined ? POWERUP_INDEX[pid] : -1;
        // Single scalar encoding type (saves features) plus binary "any-puj" flag
        out[i++] = idx >= 0 ? (idx + 1) / NUM_POWERUP_TYPES : 0;
    } else {
        out[i++] = 0;
        out[i++] = 0;
        out[i++] = 0;
        out[i++] = 0;
    }

    // Pad zeros to FEATURE_DIM exactly
    while (i < FEATURE_DIM) out[i++] = 0;
    return out;
}

// Predict ball using the same friction model as the game
function predictBall(ball, frames) {
    let x = ball.x, y = ball.y, vx = ball.vx, vy = ball.vy;
    const f = Physics.BALL_FRICTION;
    for (let i = 0; i < frames; i++) {
        vx *= f; vy *= f;
        x += vx; y += vy;
    }
    return { x, y };
}

// Decode actor logits into a usable game action. The actor outputs 5 numbers:
//   [0,1] mu_moveX, mu_moveY: continuous, squashed by tanh -> direction
//   [2]   mu_charge: continuous, squashed by sigmoid -> 0..1
//   [3]   kick_logit: discrete, sigmoid -> probability
//   [4]   pull_logit: discrete, sigmoid -> probability
//
// At training time we sample (Gaussian for continuous, Bernoulli for discrete);
// at runtime we take the deterministic mean.
function decodeActionDeterministic(self, ball, isRed, mirrorY, raw) {
    const mvx = Math.tanh(raw[0]);
    const mvy = Math.tanh(raw[1]);
    const charge = sigmoid(raw[2]);
    const kickP = sigmoid(raw[3]);
    const pullP = sigmoid(raw[4]);

    // Convert from side coords back to world
    const worldVx = isRed ? mvx : -mvx;
    const worldVy = mirrorY ? -mvy : mvy;
    return {
        moveX: worldVx,
        moveY: worldVy,
        charge: charge,
        kick: kickP > 0.5,
        pull: pullP > 0.5,
    };
}

function sigmoid(x) {
    if (x >= 0) {
        const z = Math.exp(-x);
        return 1 / (1 + z);
    } else {
        const z = Math.exp(x);
        return z / (1 + z);
    }
}

// FrameStack: tiny utility that keeps the last K encoded frames per side
// and produces the stacked input vector. Used by both env and runtime agent.
class FrameStack {
    constructor(featureDim, k) {
        this.featureDim = featureDim;
        this.k = k;
        this.frames = []; // length k, oldest first
        this.stacked = new Float32Array(featureDim * k);
    }
    // Replace all k slots with this single frame (used on reset / goal-reset)
    fill(frame) {
        this.frames = [];
        for (let i = 0; i < this.k; i++) this.frames.push(new Float32Array(frame));
        this._rebuild();
    }
    // Push a new frame, dropping the oldest
    push(frame) {
        this.frames.shift();
        this.frames.push(new Float32Array(frame));
        this._rebuild();
    }
    _rebuild() {
        const fd = this.featureDim;
        for (let i = 0; i < this.k; i++) {
            const f = this.frames[i];
            this.stacked.set(f, i * fd);
        }
    }
    get() { return this.stacked; }
}

return {
    FEATURE_DIM,
    STACK_K,
    STACKED_DIM,
    encode,
    decodeActionDeterministic,
    sigmoid,
    predictBall,
    FrameStack,
};

});
